algviz/linkedlist/linkedlistvisualizer.py

```diff
@@ -1,152 +1,152 @@
-			else:
-				center[0] += offset/2
-				if not getSize:
-					self.svg_doc.add(self.svg_doc.ellipse(center = center, r = [radius + offset/2, radius], stroke_width = '3', stroke = 'black', fill = 'white'))
-					self.draw_contents(i, self.svg_doc, center, nested = True)
-				center[0] += 2.5 *radius + offset/2#x coord of center moves right 100 + offset//2 px each iteration, we added offset earlier
-				size += 2.5*radius + offset
-		center = originalCenter#this is whatever the original center for the first node is
-		for i in range(0, len(offset_array)):#add edges
-			if i != length - 1:#dont add an edge on the last one
-				if nested:
-					center[0] += offset_array[i] / 4
-					if not getSize:
-						self.svg_doc.add(self.svg_doc.line(start = [center[0] + radius/2 + offset_array[i]/4, center[1]], end = [center[0] + (3 * radius) / 4 + offset_array[i]/4, center[1]], stroke_width = '3', stroke = 'black'))
-					center[0] += radius*1.25 + offset_array[i]/4
-				else:
-					center[0] += offset_array[i]/2
-					if not getSize:
-						self.svg_doc.add(self.svg_doc.line(start = [center[0] + radius + offset_array[i]/2, center[1]], end = [center[0] + (3 * radius) / 2 + offset_array[i]/2, center[1]], stroke_width = '3', stroke = 'black'))
-					center[0] += 2.5*radius + offset_array[i]/2
-		self.save()
-		return size
-
-	def __str__(self):
-		return self.svg_doc.tostring()
-
-class LLVisualiser(BaseVisualizer):
-	def __init__(self, svg_doc = None):
-		self.svg_doc = svg_doc
-
-	def _draw(self, ll, center = None, nested = False, getSize = False):
-		self.draw_contents(item = ll)
-
-if __name__ == "__main__":
-	ll = LinkedList()
-	ll2 = LinkedList()
-	n7 = Node(4, None)
-	n6 = Node(111, n7)
-	n8 = Node(0, n6)
-	ll2.addNode(n8)
-	n5 = Node(321, None)
-	n4 = Node(144123, n5)
-	n3 = Node(ll2, n4)
-	n2 = Node(2232, n3)
-	n1 = Node(123, n2)
-	ll.addNode(n1)
-
-
-	t = LLVisualiser()
-	t._draw(ll)
+import svgwrite
+class Node:
+	def __init__(self, value = None, nextnode = None):
+		self.value = value
+		self.nextnode = nextnode
+
+	def setValue(self, value):
+		self.value = value
+
+	def setNext(self, nextnode):
+		self.nextnode = nextnode
+
+class LinkedList:
+	def __init__(self, start = None):
+		self.start = start
+
+	def addNode(self, node):
+		current = self.start
+		if(self.start is None):
+			self.start = node
+		else:
+			while(current.nextnode is not None):
+				current = current.nextnode
+			current.nextnode = node#add the node
+
+#	def length(self):
+#		current = self.start
+#		count = 0;
+#		while(current is not None):
+#			count += 1
+#			current = current.nextnode
+#		return count
+
+	def __iter__(self):
+		self.i = self.start
+		return self
+
+	def __next__(self):
+		if self.i is None:
+			raise StopIteration
+		else:
+			item = self.i.value
+			self.i = self.i.nextnode
+			return item
+
+class BaseVisualizer:
+	def __init__(self, svg_doc = None):
+		self.svg_doc = svg_doc
+
+	def save(self):
+		if self.svg_doc is not None:
```

Comment:
In cases like this, it's nice to throw an exception in the error case instead of silently doing nothing.
```python
if self.svg_doc is None:
    raise SomeHelpfulSubclassOfError("Cannot call save without doing such-and-such first")
else:
    ...
```